Fuse Test Etag prefetcher

sauraank opened the pull request:
I have added the fuse test for checking E-tag behaviour, i.e, it is not prefetching the request from object if the object is mutated in between. In mock session it is always creating same E-Tag because `client = Arc::new(MockClient::new(client_config));`. ( new method for MockClient is adding default resulting in same e-tag always). So, I will be making that change further.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and I agree to the terms of the [Developer Certificate of Origin (DCO)](https://developercertificate.org/).

sauraank commented:
After resolving the conflicts, few of the things have changed which I need to work on to get the tests passing.

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+
+use crate::fuse_tests::TestClientBox;
+
+// test for checking either prefetching fails or read original object when object is mutated during read.
```
nit: this is a comment documenting the function, so it should be a doc comment (`///`) so that it shows up in your IDE, in documentation we generate, etc

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+    let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);
+    let mut read_buf = vec![0u8; OBJECT_SIZE];
+
+    // Make sure there's an existing directory
```
probably just delete this comment (we don't care about the directory in this test)

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+
+    let mut path = mount_point.path().join("dir/hello.txt");
+
+    // Opening the file to read and write
```
it's open read-only here, though. probably delete this comment too -- it's just repeating the code rather than explaining anything else.

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+    let mut buf = vec![0u8; read_size];
+    f.read_exact(&mut buf).expect("Should be able to read file to buf");
+
+    read_buf = vec![255u8; OBJECT_SIZE];
+    test_client.put_object("dir/hello.txt", &read_buf).unwrap();
```
explain in a comment why we're doing this (why 255? why write now?)

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+    read_buf = vec![255u8; OBJECT_SIZE];
+    test_client.put_object("dir/hello.txt", &read_buf).unwrap();
+
+    // In order to read from a position where prefetching next block is necessary
+    f.seek(std::io::SeekFrom::Start(SEEK_POS))
+        .expect("Seek position should be less than object size");
+    let mut dest_buf = vec![0u8; read_size];
+    match f.read_exact(&mut dest_buf) {
+        Ok(()) => assert_eq!(buf, dest_buf),
+        Err(err) => assert_eq!(err.raw_os_error(), Some(libc::EIO)),
+    };
```
rather than seeking, could we just keep reading until we get an error or reach end-of-file? then we can validate that no byte we successfully read is 255. that will make this test a little more resilient to the size of the prefetch window.

sauraank replied:
Like starting to read from start and keep reading the full object size? In that case I should read the object full in the first read as well. So, all the blocks will be prefetched and we will always get one of the case, i.e., data in `dest_buf `same as `buf`. We will never get the case of prefetching failing due to etag.

jamesbornholt replied:
No, not quite, I mean something like this:
```
read next 100KiB
do the new put
while !EOF {
    read next 100KiB
    assert!(read failed || read saw the old bytes);
    if read failed {
        break;
    }
}
```
Basically, do a small read from the file first (to start the prefetcher), then do the PUT, then read the rest of the file. We expect the reads to fail eventually (once we get outside the already-prefetched data), but in the meantime, they should still return the old values. If they ever return the new bytes, that's a bug.

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+
+    drop(f);
+
+    // File wont be mutated during read now to check if prefetcher reads correctly otherwise.
```
I think the actual explanation here is that re-opening the file will see the new file contents.

jamesbornholt commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,101 @@
+#[test_case(512 * 1024, 1; "first request size greater than seek position reading 1 byte")]
+#[test_case(256 * 1024, 200 * 1024; "default first request size reading till seek")]
+fn prefetch_test_etag_mock(request_size: usize, read_size: usize) {
+    // prefetch_test_etag(
+    //     crate::fuse_tests::mock_session::new,
+    //     "prefetch_etag_test",
+    //     request_size,
+    //     read_size,
+    // );
```
could we implement the etag check in the mock client too? shouldn't be very hard, and then we'll be able to have a version of this test that doesn't require S3 creds, which we should aspire to do whenever we can

sauraank replied:
Yes. We will have to generate a new e-tag when we create a mock client. Thanks.

sauraank replied:
Oh! It is failing with different etags as well in mock client. I need dive deeper on how mock client is reading.

jamesbornholt reviewed (commented):
Neat! Just one suggestion about doing this test slightly differently, and a few small things.

passaro commented on mountpoint-s3-client/src/mock_client.rs:
```diff
@@ -252,6 +269,12 @@ impl ObjectClient for MockClient {
     ) -> ObjectClientResult<Self::GetObjectResult, GetObjectError, Self::ClientError> {
         trace!(bucket, key, ?range, ?if_match, "GetObject");
 
+        if let Some(etag) = if_match {
```
Shouldn't the etag be checked after the object is found? 

I think that also means we do not need the `get_etag` method.

passaro commented on mountpoint-s3-client/src/mock_client.rs:
```diff
@@ -291,13 +314,14 @@ impl ObjectClient for MockClient {
 
         let objects = self.objects.read().unwrap();
         if let Some(object) = objects.get(key) {
+            let etag = object.etag.clone().expect("E-Tag should be set");
```
Why is it an Option then?

passaro commented on mountpoint-s3/tests/fuse_tests/mod.rs:
```diff
@@ -84,7 +87,8 @@ mod mock_session {
     impl TestClient for MockTestClient {
         fn put_object(&mut self, key: &str, value: &[u8]) -> Result<(), Box<dyn std::error::Error>> {
             let full_key = format!("{}{}", self.prefix, key);
-            self.client.add_object(&full_key, value.into());
```
Would it be better to do this in the `From` implementation for MockObject? (mock_client.rs:161)

passaro commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -85,16 +84,16 @@ where
+        read_size,
+    );
 
     prefetch_test_etag(
```
I think the pattern we are using elsewhere is to have 2 functions, see for example write_tests.rs. 

In this case, they would be: 
* `fn prefetch_test_etag_mock` 
* `fn prefetch_test_etag_s3` with `#[cfg(feature = "s3_tests")]`

That means there can be some duplication in the `test_case` attributes, but in general they do not need to be the same.

passaro commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -30,12 +30,10 @@ where
     let (mount_point, _session, mut test_client) = creator_fn(prefix, test_config);
     let mut read_buf = vec![0u8; OBJECT_SIZE];
```
Could we use a more descriptive name here? Maybe: 
```suggestion
    let original_data_buf = vec![0u8; OBJECT_SIZE];
```

passaro commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -44,6 +42,7 @@ where
     f.read_exact(&mut buf).expect("Should be able to read file to buf");
 
+    // changed the value of read_buf to distinguish it from previous data of the object.
     read_buf = vec![255u8; OBJECT_SIZE];
```
```suggestion
    let new_data_buf = vec![255u8; OBJECT_SIZE];
```

passaro replied:
Also, the assert on line 69 could compare `dest_buf` with a slice from this buffer.

passaro commented on mountpoint-s3-client/src/object_client.rs:
```diff
@@ -20,11 +22,22 @@ impl ETag {
+
+    // Creating unique etag from bytes
+    pub fn from_object_bytes(data: &[u8]) -> Self {
+        let mut hasher = Md5::new();
+        hasher.update(data);
+
+        let hash = hasher.finalize();
```
nit:  have you considered something like this:
```suggestion
        let hash = Md5::new()
            .update(data)
            .finalize();
```
may be subjective though :)

sauraank commented:
I have made test to keep prefetching until it fails or reach end of file. Also, added checking E-Tag during Mock session by adding E-Tag in Mock Client's head_object and list_object. Added the if_match parameter in get_object() of Mock Client. Please review it.

passaro commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,121 @@
+    drop(f);
+}
+
+// Prefetching of next request occurs when more than half of the current request is being read.
```
Are these comments for the test or the specific cases?

Ideally, each case description should give enough details (or they could have a specific comment if really needed).

If the comment is for the whole test, we could expand the description of `prefetch_test_etag` instead.

sauraank replied:
Like, it is generally adding why the test should behave the way they are. I can move it to the description of `prefetch_test_etag` instead.

sauraank commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,121 @@
+                if let Some(error) = err.raw_os_error() {
+                    assert_eq!(error, libc::EIO);
+                }
+                break;
```
Getting a case where the error is None. Looking into it.

sauraank replied:
Not getting the None error now.

sauraank commented on mountpoint-s3/tests/fuse_tests/prefetch_test.rs:
```diff
@@ -0,0 +1,121 @@
+
+    // Reading the file untill we keep getting the prefetched data or we get an IO error where E-Tag did not match.
+    loop {
+        seek_pos = match f.seek(std::io::SeekFrom::Start(seek_pos + read_size as u64)) {
```
I dont think so. It is seeking to the next block of read_size to keep reading untill we are not able to seek (which should be EOF)?

sauraank commented on mountpoint-s3-client/src/mock_client.rs:
```diff
@@ -291,13 +305,14 @@ impl ObjectClient for MockClient {
 
         let objects = self.objects.read().unwrap();
         if let Some(object) = objects.get(key) {
+            let etag = object.etag.clone();
```
Yes, It is. Because `object` type is `&Arc<MockObject>`. And we cannot move etag out of an Arc without cloning.

sauraank commented on mountpoint-s3-client/src/mock_client.rs:
```diff
@@ -93,7 +93,7 @@ pub struct MockObject {
     storage_class: String,
     pub last_modified: OffsetDateTime,
-    pub etag: Option<ETag>,
+    pub etag: ETag,
```
Right. Making the change.

sauraank replied:
Done. Also, made changes for other recommendations. Thanks.
